### PR TITLE
[flink] Add base support of computed column for CDC

### PR DESCRIPTION
**Pull request by yuzelin**
### Purpose

Base of #1077 and #1038 

### Tests

`MySqlComputedColumnE2ETest`
`MySqlSyncTableActionITCase#testComputedColumn`

### API and Format

New argument in mysql-sync-table:
--computed-column

### Documentation

CDC Ingestion

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Please resolve conflicts.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSchema.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import java.sql.DatabaseMetaData;
 import java.sql.ResultSet;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSchema.java:32]: We can make this class pure, without `caseSensitive`.

File: paimon-common/src/main/java/org/apache/paimon/utils/DateTimeUtils.java [hunk 2]
```diff
@@ -416,6 +416,10 @@ public static Timestamp parseTimestampData(String dateStr, int precision)
                 fromTemporalAccessor(DEFAULT_TIMESTAMP_FORMATTER.parse(dateStr), precision));
     }
 
+    public static LocalDateTime toLocalDateTime(String dateStr) {
+        return fromTemporalAccessor(DEFAULT_TIMESTAMP_FORMATTER.parse(dateStr), 0);
```
[the remark this answers is not in the document]
Comment (yuzelin) [reply paimon-common/src/main/java/org/apache/paimon/utils/DateTimeUtils.java:420]: My thought was currently we don't need the nanoseconds part. Maybe the default can be 9.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/ColumnValueComputer.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    String inputName();
+
+    /** Return {@link DataType} of computed value. */
+    DataType outputType();
+
+    /** Compute value from given input. Input and output are serialize to string. */
+    String computeValue(String input);
+
+    static ColumnValueComputer create(
+            String exprName, Map<String, DataType> typeMapping, String[] args) {
```
[the remark this answers is not in the document]
Comment (yuzelin) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/ColumnValueComputer.java:48]: But I think we need inputType and args.
inputType: `truncate` can process numeric, string. We need input type.
args: still truncate: `truncate(width, fieldReference)`. We need the arg width.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSchema.java (hunk 1 above)
Comment (yuzelin) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSchema.java:32]: Done by: https://github.com/apache/incubator-paimon/pull/1123

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/ColumnValueComputer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/ColumnValueComputer.java:48]: `String exprName, String fieldReference, DataType fieldType, Object literal`?
Is this enough? 
Expression is a very complex thing, and its parameter definitions are very complex. I don't want to touch it to the point of complexity, we only need to meet the requirements now.

File: paimon-common/src/main/java/org/apache/paimon/utils/DateTimeUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/utils/DateTimeUtils.java:420]: We can just introduce a method with precision, and use `toLocalDateTime(str, 0)`

File: docs/content/how-to/cdc-ingestion.md [hunk 4]
```diff
@@ -56,6 +57,8 @@ To use this feature through `flink run`, run the following shell command.
 * `--table` is the Paimon table name.
 * `--partition-keys` are the partition keys for Paimon table. If there are multiple partition keys, connect them with comma, for example `dt,hh,mm`.
 * `--primary-keys` are the primary keys for Paimon table. If there are multiple primary keys, connect them with comma, for example `buyer_id,seller_id`.
+* `--computed-column` are the definitions of computed columns. Supported expressions are:
```
Comment (JingsongLi) [inline docs/content/how-to/cdc-ingestion.md:60]: Document mysql field name?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncTableAction.java [hunk 5]
```diff
@@ -143,21 +167,41 @@ public void build(StreamExecutionEnvironment env) throws Exception {
 
         Identifier identifier = new Identifier(database, table);
         FileStoreTable table;
-        Schema fromMySql =
-                MySqlActionUtils.buildPaimonSchema(
-                        mySqlSchema, partitionKeys, primaryKeys, tableConfig, caseSensitive);
+        List<ComputedColumn> computedColumns = new ArrayList<>();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncTableAction.java:170]:
```
List<ComputedColumn> computedColumns = MySqlActionUtils.buildComputedColumns(
                            computedColumnArgs, mySqlSchema.typeMapping());
Schema fromMySql =
                    MySqlActionUtils.buildPaimonSchema(
                            mySqlSchema,
                            partitionKeys,
                            primaryKeys,
                            computedColumns,
                            tableConfig,
                            caseSensitive);
try {
            table = (FileStoreTable) catalog.getTable(identifier);
            checkArgument(
                    computedColumns.isEmpty(),
                    "Cannot add computed column when table already exists.");
            MySqlActionUtils.assertSchemaCompatible(table.schema(), fromMySql);
        } catch (Catalog.TableNotExistException e) {
            catalog.createTable(identifier, fromMySql, false);
            table = (FileStoreTable) catalog.getTable(identifier);
        }
```

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java [hunk 6]
```diff
@@ -262,4 +272,37 @@ private static void validateMySqlConfig(Configuration mySqlConfig) {
...
+                    String.format(
+                            "Invalid expression: %s. Please use format 'expr-name(args, ...)'.",
+                            expression));
+
+            String exprName = expression.substring(0, left);
+            String[] args = expression.substring(left + 1, right).split(",");
+            for (int i = 0; i < args.length; i++) {
+                args[i] = args[i].trim();
+            }
+            computedColumns.add(new ComputedColumn(columnName, exprName, typeMapping, args));
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java:303]: Just move the codes in `ComputedColumn` to here, and just introduce `ComputedColumn(String columnName, Expression expression)`

**Review by JingsongLi — APPROVED** [review]
LGTM

**Comment by JingsongLi** [comment]
Please merge by yourself after test passing.

**Pull request merged**